### PR TITLE
Fixed IntegerFormatStyle for values > Int64.max.

### PR DESCRIPTION
**Pull request by wadetregaskis**
If I understand the NumberFormatter implementation in the ICU library, this _should_ handle arbitrary-precision numbers, as long as they have a `description` which returns their POSIX-formatted representation, although that's not in the unit tests because it requires a working arbitrary-precision type, and I assume it's not acceptable to pull in a BigInt package for this purpose, and my efforts to make a quick-and-dirty one for testing purposes led only to tears and threats against Xcode's life.

### CODE REVIEW

**Comment by itingliu** [comment]
@swift-ci please test

**Comment by itingliu** [comment]
@swift-ci please test

**Comment by itingliu** [comment]
To be precise, ICU requires the string to be a "numeric string" as defined here: https://speleotrove.com/decimal/daconvs.html#refnumsyn

Ideally we would want to implement our own if we are to follow the spec, but realistically I doubt swift's implementation of `description` will ever change, so it'd probably work fine. Using `description` however doesn't feel right... I'm curious to hear what others think. @parkera @glessard ?

**Comment by parkera** [comment]
We've certainly been known to change descriptions for various reasons over the years, so I would vote for using a dedicated internal function to be clear what the requirements are.

**Comment by wadetregaskis** [comment]
> Ideally we would want to implement our own if we are to follow the spec, but realistically I doubt swift's implementation of `description` will ever change, so it'd probably work fine. Using `description` however doesn't feel right... I'm curious to hear what others think. @parkera @glessard ?

It's not just what the built-in integer types do for `description`, it's 3rd party types too.  My reason for doing this, for example, is because I want to use arbitrary-precision integers (via [BigInt](https://github.com/attaswift/BigInt)) while still having usable display in GUIs (because big numbers _especially_ need things like thousands separators).

There are [many 'BigInt' packages for Swift](https://swiftpackageindex.com/search?query=arbitrary+precision), and while I didn't exhaustively check them, I did look at most and they all do render the desired output from their `description` (because none of them remotely try to produce localised or "nicely" formatted results, they all do the same basic thing of just splatting out the latin decimal digits).

I did initially intend to write a BinaryInteger-to-Unicode-number-string function for this, but deferred it given:

* All my examples already produce the right output.
* I don't know what `BinaryInteger.description` intends to do as-is (e.g. maybe it's already defined to output in the desired format?  Sounds like no, though, from @parkera's comment).
* I wasn't sure how much effort it'd be to write an _efficient_ implementation.
* I couldn't conclusively locate the actual format specification, from ICU.  They refer vaguely to some other spec document, which doesn't seem to contain a clear number format specification.

If everyone's generally happy with this change otherwise, then I'm happy to dig into this last aspect and address that too.  I just didn't want to do that work pre-emptively without knowing if this fix would be welcome to begin with.

**Comment by itingliu** [comment]
Thinking more about this I think it'd be more reasonable if we implement a string following the spec of on our own 

The problem with relying on `.description` is that there's no API contract of what that returns, so the fact that it works is pure coincidence. On the contrary we have all the internal values for `BinaryInteger` and `BinaryFloatingPoint` thanks to the protocol requirement, which would allow us to implement such function reliably. 

> I wasn't sure how much effort it'd be to write an efficient implementation.

I understand this concern, but I think it's totally fine to start with something that works intuitively, and defer non-obvious optimization later if needed. I'm all for avoiding unnecessary premature optimization

> I couldn't conclusively locate the actual format specification, from ICU. They refer vaguely to some other spec document, which doesn't seem to contain a clear number format specification.

https://speleotrove.com/decimal/daconvs.html#refnumsyn this is the spec. I agree that it's unclear, but in my opinion it's not much different from how opaque `.description` is

**Comment by wadetregaskis** [comment]
Sorry for the delay - I installed Sonoma on my Macs last week and it took pretty much the entire week to repair the damage it did.  That's still a work in progress, but I can at least get Xcode to launch again, so I've picked this up.

I do have a working implementation of the additional changes, to not rely on `description`.

Unfortunately I rebased and I'm now blocked by the fact that top of tree doesn't build anymore (even before I reapply my patches).  1,513 errors, things like "'Predicate' is only available in macOS 9999 or newer" and "CircularReference".  I'll wait until those are all fixed before resuming my work.

**Comment by parkera** [comment]
> > Unfortunately I rebased and I'm now blocked by the fact that top of tree doesn't build anymore (even before I reapply my patches). 1,513 errors, things like "'Predicate' is only available in macOS 9999 or newer" and "CircularReference". I'll wait until those are all fixed before resuming my work.
> 
> Thanks for bringing this up to our attention! We're going to resolve this soon

#271

**Comment by wadetregaskis** [comment]
Does this project support 32-bit platforms (specifically, where UInt is 32-bit)?

**Comment by wadetregaskis** [comment]
Getting there.  I believe I've address all the correctness issues for positive numbers, but there's at least one remaining bug for negative.

Also, this surely won't work if UInt is 32-bit, but only because I've hard-coded the assumption of 64-bitness into the unit tests (the non-test code should work fine in any bitness - I was careful not to hard-code any assumptions there).  I don't have any way (that I know of) to actually test the project in 32-bit mode.

I also still want to optimise the output generation - it's currently using a little over twice as much memory as it has to, due to the two-phase process of first assembling the 'word' blocks in reverse order, then glueing them together.  I think I can combine those two (possibly making the code a little simpler overall).

**Comment by wadetregaskis** [comment]
@swift-ci please test

**Comment by wadetregaskis** [comment]
Close to finished now.  I still need to do a final review and clean up any rough edges or outdated comments.

I added a dependency on attaswift/BigInt in order to unit test against a real arbitrary-precision integer.  I kept that to a single patch so it's easy to exclude or revert - though that dependency is only used for testing purposes, I don't know if it's considered kosher to include it.

The unit tests still only work in 64-bit, too.  Let me know if that needs to change.

**Comment by wadetregaskis** [comment]
Okay, I think I'm done refining and reviewing it.

Let me know what you think.

**Comment by parkera** [comment]
Sorry I didn't get your comment earlier -- but yes, we do support 32-bit platforms.

**Comment by wadetregaskis** [comment]
In theory it now all works when UInt is 32-bit.  I have no way to test that, though, that I'm aware of.

**Comment by itingliu** [comment]
> I added a dependency on attaswift/BigInt in order to unit test against a real arbitrary-precision integer. I kept that to a single patch so it's easy to exclude or revert - though that dependency is only used for testing purposes, I don't know if it's considered kosher to include it.

I'd prefer we don't pull in another package. It's still yet another dependency required for building the package even if it's used for testing. Also this makes the tests dependent on the package's implementation, which complicates the tests if the implementation changes. It might make sense to have that as integration test with that specific package, but not as unit tests in swift-foundation. The tests you added are already pretty comprehensive in my opinion.

File: Package.swift
Comment (glessard) [inline Package.swift:27]: Could you get enough test coverage by using `Int128` from swift-numerics? This is something you could include. It would not be able to expose performance problems in extremely large numbers, but it should allow you to validate most of the algorithm.

Comment (wadetregaskis) [reply Package.swift:27]: `DoubleWidth` is (essentially) package-internal to swift-numerics; it's not exported.

**Comment by wadetregaskis** [comment]
> I'd prefer we don't pull in another package. It's still yet another dependency required for building the package even if it's only used for testing. Also this makes the tests dependent on the package's implementation, which complicates the tests if the implementation changes. It might make sense to have that as integration test with that specific package, but not as unit tests in swift-foundation. The tests you added are already pretty comprehensive in my opinion.

Fair enough.  How about I put those bits behind `#if canImport(BigInt)` and comment out the import (in Package.swift), so that they're there for posterity if anyone wants to recheck with them, but they won't incur any overheads (dependent package fetching etc) for everyone, by default?

The BigInt-based tests caught most of the bugs, as I was implementing things.  There's not much useful surface area in the built-in integer types since nothing interesting happens until at least Int64.max + 1.

It's a shame that while the Swift Stdlib actually contains an Int128 type, it just doesn't export it.  Int128 wouldn't fully cover the testing desires, but it would be a lot better than just UInt64 and smaller.

File: Tests/FoundationInternationalizationTests/Formatting/Number/BinaryInteger+FormatStyleTests.swift
Comment (glessard) [inline Tests/FoundationInternationalizationTests/Formatting/Number/BinaryInteger+FormatStyleTests.swift:1]: The copyright should be assigned to the Swift project. While we're at it, updating the copyright years of the files being modified would be welcome.

Comment (wadetregaskis) [reply Tests/FoundationInternationalizationTests/Formatting/Number/BinaryInteger+FormatStyleTests.swift]: Done.

**Comment by wadetregaskis** [comment]
In addition to addressing feedback, I did some more proof-reading and improved some of the documentation.

I also realised it's possible to detect incorrect implementations of quotientAndRemainder(dividingBy:), and that by doing so it slightly simplifies the code anyway, so I added that in.  I know it's possible to get quotientAndRemainder(dividingBy:) wrong because when I implemented it my first time I did exactly what this checks for (using T-division instead of F-division, because the former is what I'd call the standard method, mathematically).

I was also reminded that `assert()` without a bespoke message gives pretty useless output, so I added detailed error messages to all the asserts.

File: Package.swift
Comment (glessard) [reply Package.swift]: We could copy the source for DoubleWidth into the test utilities, until such time as we have something either in the stdlib or in a package we are willing to import.

**Comment by itingliu** [comment]
@swift-ci please test

File: Tests/FoundationInternationalizationTests/Formatting/Number/BinaryInteger+FormatStyleTests.swift
Comment (glessard) [inline Tests/FoundationInternationalizationTests/Formatting/Number/BinaryInteger+FormatStyleTests.swift:44]: Typo:
```suggestion
    func testNumericStringRepresentation_builtinIntegersAroundDecimalMagnitude() throws {
```

File: Sources/FoundationInternationalization/Formatting/Number/BinaryInteger+FormatStyle.swift
Comment (glessard) [inline Sources/FoundationInternationalization/Formatting/Number/BinaryInteger+FormatStyle.swift:185]: Nit: this case has already been handled in the caller, so the return statement is dead code in practice. Could be replaced by an assertion.

**Review by glessard — APPROVED** [review]
This is a great contribution, thank you! Since this particular kind of formatting is independent of locale, it would be interesting to discuss whether a `String` initializer that provides the numeric string representation of a `BinaryInteger` belongs in the standard library.

File: Sources/FoundationInternationalization/Formatting/Number/BinaryInteger+FormatStyle.swift
Comment (iCharlesHu) [inline Sources/FoundationInternationalization/Formatting/Number/BinaryInteger+FormatStyle.swift:184]: Since it's a implementation detail this method should probably be `private`.

File: Tests/FoundationInternationalizationTests/Formatting/Number/BinaryInteger+FormatStyleTests.swift
Comment (iCharlesHu) [inline Tests/FoundationInternationalizationTests/Formatting/Number/BinaryInteger+FormatStyleTests.swift:19]: Instead of using `BigInt`, we should wait until `Decimal` is implemented and use it instead: https://github.com/apple/swift-foundation/issues/236

File: Sources/FoundationInternationalization/Formatting/Number/BinaryInteger+FormatStyle.swift
Comment (iCharlesHu) [reply Sources/FoundationInternationalization/Formatting/Number/BinaryInteger+FormatStyle.swift:52]: Also, could you break down the comments into multiple lines? I know we don't have an official style guide yet, but these comments will definite exceed that limit.

Comment (wadetregaskis) [reply Sources/FoundationInternationalization/Formatting/Number/BinaryInteger+FormatStyle.swift:52]: I can.  What is the line length limit?

File: Sources/FoundationInternationalization/Formatting/Number/BinaryInteger+FormatStyle.swift
Comment (wadetregaskis) [reply Sources/FoundationInternationalization/Formatting/Number/BinaryInteger+FormatStyle.swift:59]: Perhaps; I'll make it so.

It should never trip as long as `UInt(exactly:)` and the implementation of the concrete type's `words` property aren't buggy.  If either is, then it's likely that the result of this method will be wrong anyway.  So the benefit of the check is probably just a faster and clearer failure.

But the downside of the precondition check seems quite small - slight increase in code size and a [hopefully very small] runtime cost.

**Comment by wadetregaskis** [comment]
> This is a great contribution, thank you! Since this particular kind of formatting is independent of locale, it would be interesting to discuss whether a `String` initializer that provides the numeric string representation of a `BinaryInteger` belongs in the standard library.

I think that's a good direction.  As I noted in the code, another benefit of moving this into the standard library is that various `description`s can use it, eliminating duplicate code (I haven't looked at how `description` is implemented for the standard numeric types, but it's likely _very_ similar to what I've implemented here).

File: Tests/FoundationInternationalizationTests/Formatting/Number/BinaryInteger+FormatStyleTests.swift
Comment (wadetregaskis) [reply Tests/FoundationInternationalizationTests/Formatting/Number/BinaryInteger+FormatStyleTests.swift:19]: To be clear, are you saying to remove this code entirely?  It's up to you, but FWIW I feel it's useful to keep accessible - e.g. while addressing feedback I locally re-enabled the BigInt package dependency in order to re-test things.

File: Package.swift
Comment (wadetregaskis) [reply Package.swift]: To clarify, I don't have strong feelings on this, I'd just like to be sure it's a fully informed decision.  Copy-pasting `DoubleWidth` is mechanically easy but it's against general best practice.

[I did suggest putting `DoubleWidth` into its own package](https://forums.swift.org/t/status-check-int128-uint128/67694/17), so that it can be more cleanly reused, but @stephentyrone discouraged it.

Having use of 128-bit integers would improve the unit test's effective coverage significantly.  It's not enough to provide full confidence, since some bugs by nature are only hit with _much_ larger numbers, but it's fairly close.

**Comment by tbkka** [comment]
> ... benefit of moving this into the standard library is that various descriptions can use it, eliminating duplicate code ...

I think it would be useful to have a standard fallback implementation that can handle user-defined types.

I'd want to do some performance profiling before adopting it for the standard numeric types, as the `description` methods are used in some performance-critical code (such as JSON formatting).

**Comment by wadetregaskis** [comment]
> I'd want to do some performance profiling before adopting it for the standard numeric types, as the `description` methods are used in some performance-critical code (such as JSON formatting).

Yes, that would be wise.

I didn't do any comparative performance analysis with this implementation.  I did take some pains to _try_ to make it quite fast - the code could be a lot simpler otherwise - but that's all.  A significant factor in its performance is how well the compiler optimises it, of course, but an even bigger one is how the actual type in question performs as a `BinaryInteger` (particularly re. `quotientAndRemainder(dividingBy:)`).  And that raises the question of which BigInt implementation you use for reference, etc.  So, potentially quite an undertaking to really evaluate robustly.

For built-in types, this will likely be _slower_ than dedicated implementations (as currently written) because of things like `decimalDigitsAndMagnitudePerWord` which are fairly fast but can be just hard-coded constants in non-generic implementations.

File: Tests/FoundationInternationalizationTests/Formatting/Number/BinaryInteger+FormatStyleTests.swift
Comment (wadetregaskis) [reply Tests/FoundationInternationalizationTests/Formatting/Number/BinaryInteger+FormatStyleTests.swift]: Well spotted!  Fixed.

File: Sources/FoundationInternationalization/Formatting/Number/BinaryInteger+FormatStyle.swift
Comment (wadetregaskis) [reply Sources/FoundationInternationalization/Formatting/Number/BinaryInteger+FormatStyle.swift]: Done.

File: Package.swift
Comment (wadetregaskis) [reply Package.swift]: Done.

File: Sources/FoundationInternationalization/Formatting/Number/BinaryInteger+FormatStyle.swift
Comment (wadetregaskis) [reply Sources/FoundationInternationalization/Formatting/Number/BinaryInteger+FormatStyle.swift]: Done.

File: Sources/FoundationInternationalization/Formatting/Number/BinaryInteger+FormatStyle.swift
Comment (wadetregaskis) [reply Sources/FoundationInternationalization/Formatting/Number/BinaryInteger+FormatStyle.swift]: Done (well, `fileprivate` by necessity, but yes).

File: Sources/FoundationInternationalization/Formatting/Number/BinaryInteger+FormatStyle.swift
Comment (parkera) [reply Sources/FoundationInternationalization/Formatting/Number/BinaryInteger+FormatStyle.swift]: We don't have a line length limit in this project, preferring to let users wrap it to their own preference in their editors.

Comment (parkera) [reply Sources/FoundationInternationalization/Formatting/Number/BinaryInteger+FormatStyle.swift]: That said - I agree that we should keep header doc and implementation comments split up.

Comment (wadetregaskis) [reply Sources/FoundationInternationalization/Formatting/Number/BinaryInteger+FormatStyle.swift]: So I think this is resolved, then?  I didn't hard-wrap anything, but I moved the couple of paragraphs in question into the function body as regular comments.

File: Sources/FoundationInternationalization/Formatting/Number/BinaryInteger+FormatStyle.swift
Comment (oscbyspro) [inline Sources/FoundationInternationalization/Formatting/Number/BinaryInteger+FormatStyle.swift:82]: The bit-width of fixed-width integers is not normalized, so `remainder.bitWidth` may be fixed to a value larger than `UInt.bitWidth` (see `Int256`).

**Review by oscbyspro — CHANGES_REQUESTED** [review]
Hi, I saw this pull request mentioned on the Swift forums and wanted to try it out. While doing so, I noticed that a debug assertion fails for fixed-width integers wider than `UInt` (see comment). Without this assertion, the rest works with my double-width integer in [Numberick](https://github.com/oscbyspro/Numberick).

File: Sources/FoundationInternationalization/Formatting/Number/BinaryInteger+FormatStyle.swift
Comment (wadetregaskis) [reply Sources/FoundationInternationalization/Formatting/Number/BinaryInteger+FormatStyle.swift:82]: Thanks for catching that.

This would presumably also make Int64 & UInt64 fail in 32-bit mode.

It's very unfortunate that it's documented to behave that way; it seems to render that property useless.  The full width of fixed-width integers is already provided by the type-level property `bitWidth`.

The claimed bit width is relied upon for more than just that assert; I'll have to think a bit about how to work around this problem…

Comment (wadetregaskis) [reply Sources/FoundationInternationalization/Formatting/Number/BinaryInteger+FormatStyle.swift:82]: …which, it turns out, is not easy.  I'm still trying to find a way; no luck today, I'll have to pick this up again some other day.

Mathematically it's trivial, but because of the structure of the `BinaryInteger` and associated types (in particular that `Magnitude` is merely `Numeric`, which is virtually useless) it is hard to actually implement the necessary functionality.

Though in looking at this I realised the size estimation, based on bit width, was technically wrong anyway.  Even when `bitWidth` returns what I was expecting.  I don't _think_ it could ever actually cause an off-by-one bug because I think it would require a power of ten and a power of two to be equal (which is impossible), but still, it'd be nice to eliminate that possibility as a matter of principle.  I'm still looking into that, too.

File: Sources/FoundationInternationalization/Formatting/Number/BinaryInteger+FormatStyle.swift
Comment (wadetregaskis) [reply Sources/FoundationInternationalization/Formatting/Number/BinaryInteger+FormatStyle.swift:240]: Actually, no, reverted.  We'd both overlooked the fact that it _can_ be called on zero because it's used for the individual words of the BinaryInteger (from the generic `numericStringRepresentation` directly, not the UInt specialisation!), any of which can actually be zero.

File: Sources/FoundationInternationalization/Formatting/Number/BinaryInteger+FormatStyle.swift
Comment (wadetregaskis) [reply Sources/FoundationInternationalization/Formatting/Number/BinaryInteger+FormatStyle.swift]: Okay, life found a way.  The end result isn't that complicated, actually, and might even be more efficient than before since we can take advantage of the fact that we only care about the _magnitude_, really, as that's what we're essentially converting to decimal digits.  So a lot of the complexities of two's complement representation can be side-stepped.

A positive coincidental side-effect of this is that the algorithm no longer relies on the `bitWidth` instance property at all (though still the type-level version for UInt, which is fine).  That lack of reliance is is good because it turns out some 3rd party implementations of `BinaryInteger` don't implement it correctly anyway (e.g. https://github.com/attaswift/BigInt/issues/115).

Comment (wadetregaskis) [reply Sources/FoundationInternationalization/Formatting/Number/BinaryInteger+FormatStyle.swift]: Oh, and also, [attaswift/BigInt](https://github.com/attaswift/BigInt) conforms both its types (`BigInt` and `BigUInt`) to `FixedWidthInteger` even though they're not fixed width.  Perhaps just to make it more convenient to use (I actually [ab]use this in the unit tests here, to use the `init` method that takes a string with configurable radix)?  So my initial attempts to special-case `bitWidth` interpretation for non-`FixedWidthInteger` `BinaryInteger`s was doomed to fail anyway (or, at least, to be irrelevant).

**Comment by wadetregaskis** [comment]
I added unit tests using Numberick (https://github.com/oscbyspro/Numberick) in the same vein as the ones using BigInt, which found one additional bug.  Mostly the code doesn't care about fixed-width vs flexible-width but there is an underlying difference as the former may zero-pad `words` to the expected length, whereas the latter doesn't have any reason to (but also may have redundant, zero-padding words, I suppose - but attaswift/BigInt doesn't).

**Comment by wadetregaskis** [comment]
@iCharlesHu & @glessard, I think this is just waiting on your clarifications regarding your earlier comments.  Can you take a quick look, please?

**Comment by wadetregaskis** [comment]
So… should I drop this PR?  It seems like it's not going anywhere. 😞

I'm not sure how best to put this; I don't want to offend… this was my first attempt to contribute to an Apple open source project, and I had quite some enthusiasm towards some further contributions, but now I think that would be a bad idea.

This is fundamentally a trivially bug fix, for something that was _really_ obviously broken for a long time.  It ended up being many days of work for me - which would have been fine, it was educational at least, _iff_ it actually went through.  Doing all that work to no actual avail is disheartening.

**Review by itingliu — APPROVED** [review]
Sorry @wadetregaskis I was absorbed in my own work and this just slipped my mind. This looks good to merge to me. I'm going to kick off another CI run and then I'll merge it

If @iCharlesHu or @glessard has further comments we can always address them in a follow up PR

**Comment by itingliu** [comment]
@swift-ci please test

**Comment by parkera** [comment]
Thank you so much @wadetregaskis! I really appreciate this contribution and the detailed work you put into it.

**Comment by stephentyrone** [comment]
Yeah, I don't see anyway for this change to manifest that error.

**Pull request merged**
